File: src/main/java/com/kamikazejam/datastore/connections/storage/mongo/MongoTransactionHelper.java
package com.kamikazejam.datastore.connections.storage.mongo;

import com.google.common.base.Preconditions;
import com.kamikazejam.datastore.DataStoreSource;
import com.kamikazejam.datastore.base.Cache;
import com.kamikazejam.datastore.base.Store;
import com.kamikazejam.datastore.connections.storage.exception.TransactionRetryLimitExceededException;
import com.kamikazejam.datastore.util.DataStoreFileLogger;
import com.kamikazejam.datastore.util.JacksonUtil;
import com.mongodb.MongoCommandException;
import com.mongodb.client.ClientSession;
import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.model.Filters;
import com.mongodb.client.result.UpdateResult;
import org.bson.Document;
import org.jetbrains.annotations.NotNull;

import java.util.Random;
import java.util.function.Consumer;

import static com.mongodb.client.model.Filters.eq;

@SuppressWarnings("UnusedReturnValue")
class MongoTransactionHelper {
    private static final Random RANDOM = new Random();
    private static final int DEFAULT_MAX_RETRIES = 15;
    private static final int BASE_BACKOFF_MS = 50;      // Start with 50ms delay
    private static final int LINEAR_BACKOFF_MS = 20;    // Increase 20ms per attempt
    private static final int WRITE_CONFLICT_ERROR = 112;

    /**
     * Execute a MongoDB document update with retries and version checking
     * @param mongoClient The MongoDB client
     * @param collection The MongoDB collection
     * @param cache The cache containing the document
     * @param originalStore The original store to update
     * @param updateFunction The function to apply updates
     * @return Whether the update was successful
     */
    protected static <K, X extends Store<X, K>> boolean executeUpdate(
            @NotNull MongoClient mongoClient,
            @NotNull MongoCollection<Document> collection,
            @NotNull Cache<K, X> cache,
            @NotNull X originalStore,
            @NotNull Consumer<X> updateFunction
    ) {
        Preconditions.checkNotNull(mongoClient, "MongoClient cannot be null");
        Preconditions.checkNotNull(collection, "Collection cannot be null");
        Preconditions.checkNotNull(cache, "Cache cannot be null");
        Preconditions.checkNotNull(originalStore, "Store cannot be null");
        Preconditions.checkNotNull(updateFunction, "Update function cannot be null");

        try {
            // Create working copy that will be updated on each attempt
            X baseCopy = JacksonUtil.deepCopy(originalStore);
            executeUpdateInternal(mongoClient, collection, cache, originalStore, baseCopy, updateFunction, 0);
            return true;
        }catch (TransactionRetryLimitExceededException e){
            DataStoreFileLogger.warn("Failed to execute MongoDB update in " + DEFAULT_MAX_RETRIES + " attempts");
            return false;
        }
    }

    // If no error is thrown, this method succeeded
    private static <K, X extends Store<X, K>> void executeUpdateInternal(
            @NotNull MongoClient mongoClient,
            @NotNull MongoCollection<Document> collection,
            @NotNull Cache<K, X> cache,
            @NotNull X originalStore,
            @NotNull X baseCopy,
            @NotNull Consumer<X> updateFunction,
            int currentAttempt
    ) throws TransactionRetryLimitExceededException {
        // Quit if we've run out of attempts
        if (currentAttempt >= DEFAULT_MAX_RETRIES) {
            throw new TransactionRetryLimitExceededException("Failed to execute update after " + DEFAULT_MAX_RETRIES + " attempts.");
        }

        // Apply exponential backoff if this isn't our first attempt
        if (currentAttempt > 0) {
            applyBackoff(currentAttempt);
        }

        try (ClientSession session = mongoClient.startSession()) {
            session.startTransaction();
            boolean committed = false;

            try {
                // Clone the base copy for this attempt
                X workingCopy = JacksonUtil.deepCopy(baseCopy);
                workingCopy.setReadOnly(false);

                // Fetch Version prior to updates
                long currentVersion = workingCopy.getVersion().get();

                // Apply updates to the copy
                updateFunction.accept(workingCopy);
                // Increment version (Optimistic Versioning)
                workingCopy.getVersion().set(currentVersion + 1);

                final String id = cache.keyToString(workingCopy.getId());
                Document doc = JacksonUtil.serializeToDocument(workingCopy);

                UpdateResult result = collection.replaceOne(
                        session,
                        Filters.and(
                                // These two filters act as a sort of compare-and-swap mechanic
                                //  inside of this mongo transaction, if these are not met then
                                //  the transaction will fail and we will need to retry.
                                eq("_id", id),
                                eq("version", currentVersion)
                        ),
                        doc
                );

                // If no documents were modified, then the compare-and-swap failed, we must retry
                if (result.getModifiedCount() == 0) {
                    DataStoreSource.getColorLogger().debug("Failed to update Store in MongoDB Layer (Could not find document with id: '" + id + "' and version: " + currentVersion + ")");

                    // If update failed, fetch current version
                    Document currentDoc = collection.find(session).filter(eq("_id", id)).first();
                    if (currentDoc == null) {
                        throw new RuntimeException("Entity not found");
                    }

                    // Update our working copy with latest version and retry
                    baseCopy = JacksonUtil.deserializeFromDocument(cache.getStoreClass(), currentDoc);
                    executeUpdateInternal(mongoClient, collection, cache, originalStore, baseCopy, updateFunction, currentAttempt + 1);
                    return;
                }

                // Success - update the cached store from our working copy
                workingCopy.setReadOnly(true);
                originalStore.setReadOnly(false);
                cache.updateStoreFromNewer(originalStore, workingCopy);
                cache.cache(originalStore);
                originalStore.setReadOnly(true);

                session.commitTransaction();
                committed = true;

            } catch (TransactionRetryLimitExceededException t) {
                // re-throw to bubble up
                throw t;
            } catch (MongoCommandException mE) {
                if (isWriteConflict(mE)) {
                    logWriteConflict(currentAttempt);
                    // For write conflicts, retry with same working copy
                    executeUpdateInternal(mongoClient, collection, cache, originalStore, baseCopy, updateFunction, currentAttempt + 1);
                    return;
                }
                throw mE;
            } catch (Exception e) {
                DataStoreFileLogger.warn("Failed to execute MongoDB update", e);
                executeUpdateInternal(mongoClient, collection, cache, originalStore, baseCopy, updateFunction, currentAttempt + 1);
            } finally {
                if (!committed) {
                    session.abortTransaction();
                }
            }
        }
    }

    // Applies linear backoff to the current thread + some random jitter
    // Exponential backoff was way too slow, increasing the milliseconds far too fast
    // Linear backoff was selected to be more predictable and less extreme
    private static void applyBackoff(long attempt) {
        try {
            long linearBackoff = BASE_BACKOFF_MS + (LINEAR_BACKOFF_MS * attempt);
            // add the jitter (+- 25%)
            long half = linearBackoff / 2;
            long jitter = RANDOM.nextLong(half) - (half / 2);

            long backoff = linearBackoff + jitter;
            Thread.sleep(backoff);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new RuntimeException("Operation interrupted during backoff", e);
        }
    }

    private static boolean isWriteConflict(MongoCommandException e) {
        return e.getErrorCode() == WRITE_CONFLICT_ERROR;
    }

    private static void logWriteConflict(int currentAttempt) {
        DataStoreSource.getColorLogger().debug(
            "Write conflict detected, attempt " + (currentAttempt + 1) + " of " + DEFAULT_MAX_RETRIES
        );
    }
}

File: src/main/kotlin/com/kamikazejam/datastore/connections/storage/mongo/MongoTransactionHelper.kt
package com.kamikazejam.datastore.connections.storage.mongo

import com.google.common.base.Preconditions
import com.kamikazejam.datastore.DataStoreSource
import com.kamikazejam.datastore.base.Collection
import com.kamikazejam.datastore.base.exception.update.TransactionRetryLimitExceededException
import com.kamikazejam.datastore.base.exception.update.UpdateException
import com.kamikazejam.datastore.base.metrics.MetricsListener
import com.kamikazejam.datastore.base.serialization.SerializationUtil.getSerialNameForID
import com.kamikazejam.datastore.base.serialization.SerializationUtil.getSerialNameForVersion
import com.kamikazejam.datastore.store.Store
import com.kamikazejam.datastore.util.DataStoreFileLogger
import com.mongodb.MongoCommandException
import com.mongodb.client.model.Filters
import com.mongodb.kotlin.client.coroutine.ClientSession
import com.mongodb.kotlin.client.coroutine.MongoClient
import com.mongodb.kotlin.client.coroutine.MongoCollection
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.firstOrNull
import java.util.*
import java.util.logging.Level

@Suppress("MemberVisibilityCanBePrivate")
object MongoTransactionHelper {
    private val RANDOM = Random()
    var DEFAULT_MAX_RETRIES: Int = 50 // very gracious amount of retries (since we have backoff)
    var LOG_WRITE_CONFLICT_AFTER_ATTEMPT: Int = 10 // only log after this many attempts
    private const val WRITE_CONFLICT_ERROR = 112

    // Minimum and maximum backoff values to prevent extremes
    var MIN_BACKOFF_MS: Long = 50
    var MAX_BACKOFF_MS: Long = 2000
    var PING_MULTIPLIER: Double = 2.0 // Base multiplier for ping time
    var ATTEMPT_MULTIPLIER: Double = 1.5 // How much to increase per attempt

    /**
     * Execute a MongoDB document update with retries and version checking
     * @throws UpdateException for fail states
     */
    @Throws(UpdateException::class)
    suspend fun <K : Any, X : Store<X, K>> executeUpdate(
        mongoClient: MongoClient,
        mongoColl: MongoCollection<X>,
        collection: Collection<K, X>,
        store: X,
        updateFunction: (X) -> X
    ): X {
        Preconditions.checkNotNull(mongoClient, "MongoClient cannot be null")
        Preconditions.checkNotNull(mongoColl, "MongoDB Collection cannot be null")
        Preconditions.checkNotNull(collection, "Collection cannot be null")
        Preconditions.checkNotNull(store, "Store cannot be null")
        Preconditions.checkNotNull(updateFunction, "Update function cannot be null")

        // Will either return the store, or throw an UpdateException
        val updatedStore: X = retryExecutionHelper(
            mongoClient,
            mongoColl,
            collection,
            store,
            updateFunction,
            0
        )

        // Ensure our local cache is updated with the new updated store
        collection.updateStoreFromNewer(store, updatedStore)
        return updatedStore
    }

    // If no error is thrown, this method succeeded
    @Throws(UpdateException::class)
    private suspend fun <K : Any, X : Store<X, K>> retryExecutionHelper(
        mongoClient: MongoClient,
        mongoColl: MongoCollection<X>,
        collection: Collection<K, X>,
        store: X,
        updateFunction: (X) -> X,
        currentAttempt: Int
    ): X {
        // Quit if we've run out of attempts
        if (currentAttempt >= DEFAULT_MAX_RETRIES) {
            DataStoreSource.metricsListeners.forEach(MetricsListener::onUpdateTransactionLimitReached)
            throw TransactionRetryLimitExceededException("Failed to execute update after $DEFAULT_MAX_RETRIES attempts.")
        }

        // Apply exponential backoff if this isn't our first attempt
        if (currentAttempt > 0) {
            applyBackoff(currentAttempt.toLong())
        }

        // Log the Metric
        DataStoreSource.metricsListeners.forEach(MetricsListener::onTryUpdateTransaction)

        mongoClient.startSession().use { session ->
            session.startTransaction()
            var sessionResolved = false
            try {
                // Fetch Version prior to updates
                val result = processTransaction(mongoColl, session, collection, store, updateFunction)

                // Handle Fail State
                if (result.dbStore != null) {
                    session.abortTransaction()
                    sessionResolved = true
                    // Retry with the new database store
                    return retryExecutionHelper(
                        mongoClient,
                        mongoColl,
                        collection,
                        result.dbStore,
                        updateFunction,
                        currentAttempt + 1
                    )
                }

                sessionResolved = true // resolve before call, in case it fails partial
                // Success - return true
                session.commitTransaction()
                return checkNotNull(result.store)
            } catch (uE: UpdateException) {
                throw uE
            } catch (mE: MongoCommandException) {
                if (isWriteConflict(mE)) {
                    logWriteConflict(currentAttempt, mE, collection, store)
                    // For write conflicts, retry with same working copy
                    return retryExecutionHelper(
                        mongoClient,
                        mongoColl,
                        collection,
                        store,
                        updateFunction,
                        currentAttempt + 1
                    )
                }
                throw UpdateException("Failed to execute MongoDB update", mE)
            } catch (e: Exception) {
                DataStoreFileLogger.warn("Failed to execute MongoDB update", e)
                throw UpdateException("Failed to execute MongoDB update", e)
            } finally {
                if (!sessionResolved) {
                    session.abortTransaction()
                }
            }
        }
    }

    private suspend fun <K : Any, X : Store<X, K>> processTransaction(
        mongoColl: MongoCollection<X>,
        session: ClientSession,
        collection: Collection<K, X>,
        store: X,
        updateFunction: (X) -> X,
    ): TransactionResult<K, X> {
        val currentVersion: Long = store.version
        val nextVersion = currentVersion + 1
        val id: String = collection.keyToString(store.id)

        // Apply the Update Function
        //  which MUST (by convention) return a new instance (like a data class copy)
        // Also have the Store class create a copy with the desired version
        val updatedStore: X = updateFunction(store).copyHelper(nextVersion)
        if (updatedStore === store) {
            throw IllegalArgumentException("Update function must return a new instance of the store (like a data class copy)")
        }

        // Validate ID Property
        if (id != collection.keyToString(updatedStore.id)) {
            throw IllegalArgumentException("Updated store failed copy id check! Was: ${updatedStore.id}, Expected: ${store.id}")
        }

        // Validate Incremented Version (Optimistic Versioning)
        if (updatedStore.version != nextVersion) {
            throw IllegalArgumentException("Updated store failed copy version check! Was: ${updatedStore.version}, Expected: $nextVersion")
        }

        val idField = getSerialNameForID(collection)
        val verField = getSerialNameForVersion(collection)
        val result = mongoColl.replaceOne(
            session,
            Filters.and(
                // These two filters act as a sort of compare-and-swap mechanic
                //  inside of this mongo transaction, if these are not met then
                //  the transaction will fail and we will need to retry.
                Filters.eq(idField, id),
                Filters.eq(verField, currentVersion),
            ),
            updatedStore
        )

        // If no documents were modified, then the compare-and-swap failed, we must retry
        if (result.modifiedCount == 0L) {
            DataStoreSource.colorLogger.debug("Failed to update Store in MongoDB Layer (Could not find document with id: '$id' and version: $currentVersion)")

            // If update failed, fetch current version
            val databaseStore: X = mongoColl.find(session).filter(
                Filters.eq(idField, id)
            ).firstOrNull() ?: throw RuntimeException("Entity not found for collection: ${collection.name}, id: $idField -> $id")

            // Update our working copy with latest version and retry
            return TransactionResult(null, databaseStore)
        }

        // Success! Return a successful result
        return TransactionResult(updatedStore, null)
    }

    // Applies linear backoff to the current thread + some random jitter
    // Exponential backoff was way too slow, increasing the milliseconds far too fast
    // Linear backoff was selected to be more predictable and less extreme
    private suspend fun applyBackoff(attempt: Long) {
        // Divide by 2 since this method fetches the round trip time
        val pingNanos = DataStoreSource.storageService.averagePingNanos / 2
        // Convert ping from nanos to ms and apply base multiplier
        val basePingMs = (pingNanos / 1000000) * PING_MULTIPLIER.toLong()


        // Calculate backoff with attempt scaling
        var backoffMs = basePingMs + (basePingMs * ATTEMPT_MULTIPLIER * attempt).toLong()

        // Clamp the value between min and max
        backoffMs = backoffMs.coerceAtLeast(MIN_BACKOFF_MS).coerceAtMost(MAX_BACKOFF_MS)

        // Add jitter (±25%)
        val half = backoffMs / 4 // 25% of total
        val jitter = RANDOM.nextLong(-half, half)

        delay(backoffMs + jitter)
    }

    private fun isWriteConflict(e: MongoCommandException): Boolean {
        return e.errorCode == WRITE_CONFLICT_ERROR
    }


    private fun <K : Any, X : Store<X, K>> logWriteConflict(
        currentAttempt: Int,
        mE: MongoCommandException,
        collection: Collection<K, X>,
        store: X,
    ) {
        if (currentAttempt < LOG_WRITE_CONFLICT_AFTER_ATTEMPT) return

        val msg = "Write Conflict, attempt " + (currentAttempt + 1) + " of " + DEFAULT_MAX_RETRIES + " for coll " + collection.name + " with id " + collection.keyToString(store.id)
        DataStoreFileLogger.log(
            msg,
            mE,
            DataStoreFileLogger.randomWriteExceptionFile,
            Level.INFO
        )
    }
}
